torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -756,7 +757,26 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
   std::vector<at::cuda::CUDAStream> streamVal;
   streamVal.reserve(devices.size());
 
-  // Create the NCCL communicators for each GPU
+  // [Group Start/End Note] This is used to ensure that nccl communicator will be created
+  // before communication primitives are called. Let's look at this example:
+  // Using the batch_isend_irecv to send a tensor to a target process. On the sender side,
+  // the corresponding underlying NCCL calls will look like
+  //   ncclGroupStart() // This is in batch_isend_irecv
+  //   ncclGroupStart() // This is [Note 1]
+  //   ncclCommInitRank() // Inside NCCLCOmm::create
+  //   ncclSend()
+  //   ncclGroupEnd() // This is [Note 2]
+  //   ncclGroupEnd() // This is in batch_isend_irecv
+  // With this pattern, the nccl communicator will be created in the last ncclGroupEnd
+  // which means when ncclSend is processed, the passed communicator argument is NULL which will
+  // lead to runtime error. So we need to "close" all active nccl groups to ensure
+  // nccl communicator is actually created before encountering any communication calls.
+  // This is why we need the following for loop.
+  for (size_t i = 0; i < ncclActiveGroupCounter_; ++i) {
+    C10D_NCCL_CHECK(ncclGroupEnd());
+  }
```

Comment:
Instead of doing this, can we do this (which might be simpler):

1. In `_group_manager` just increment/decrement `ncclActiveGroupCounter_` via groupStart/groupEnd, but don't actually call `ncclGroupEnd` or `ncclGroupStart`.
2. In `pointTopoint`, we add the appropriate number of `ncclGroupStart` and `ncclGroupEnd` calls around the appropriate kernel?

This will probably work for our current use case, but not sure if we plan to use groupStart and groupEnd as a general API in the future. If so, this probably won't work since groupStart and groupEnd might be used with collectives and other ops too.

Reply:
ncclGroupEnd needs to be called by the last Op in the op_list, we will need to maintain another counter to indicate end of the list. So existing design with one counter might be easier.